design/14_tinkerbell_crd_refactor/proposal.md

```diff
@@ -0,0 +1,462 @@
+1. Userdata can be specified in both `.Spec.UserData` and `.Spec.Instance.Userdata`. `.Spec.Instance.Userdata` is unused.
+1. `.Spec.TinkVersion` has no functional use.
+1. `.Spec.Resources` was intended for use in CAPT as part of its Hardware selection algorithm but is yet to be implemented.
+1. `.Spec.Interfaces[].Netboot.{AllowPXE,AllowWorkflow}` and `.Spec.Metadata.Instance.AlwaysPxe` are seemingly related but reside on different objects and how they impact eachother is unclear.
+1. `.Spec.Metadata.Custom` defines specific fields that are related to other parts of Hardware.
+1. `.Status.State`, `.Spec.Metadata.Instance.State` and `.Spec.Metadata.State` have unclear semantics. The `.Spec` state fields impact the machine provisioning process but nothing in the core Tinkerbell stack sets their values. `.Status.State` is unused.
+
+**Template**
+
+1. Template defines a single field, `.Spec.Data`. The format of `.Spec.Data` is entirely ambiguous requiring the user to understand implementation detail. In summary, `.Spec.Data` is composed of a list of tasks that can run on different machines; a task is composed of a list of actions that perform a function such as streaming a raw image. The multi-machine capability has no known use-cases.
+1. The Template objects `.Status` field is unused.
+
+**Workflow**
+
+1. The `.Spec.HardwareMap` historically defines a template value used to render a tasks `WorkerAddr`. The `WorkerAddr` should be the MAC of the machine that should run the task. This creates a hard to understand relationship between Workflow and Template that users must understand to successfully execute Workflows.
+1. The `.Spec.GlobalTimeout` is unused and its origin is unclear (status fields are typically populated by Kubernetes controllers to build understanding of the current object state).
+1. Actions leverage the `WorkflowState` type that is intended to describe the overall state of the Workflow.
+
+Users resort to Q&A in the Tinkerbell Slack to determine what fields are required and how tweaking them impacts the system. The CRDs should be simple enough and sufficiently documented to aid users in understanding how they can manipulate the system.
+
+## Goals/Non-goals
+
+**Goals**
+
+- To de-duplicate Tink custom resource definition fields and data structures.
+- To provide clear behavioral expectations when manipulating custom resources.
+- To remove obsolete fields and data structures.
+
+**Non-goals**
+
+- To change the existing relationship between Tink and Rufio.
+- To introduce additional object status data typically found on the `.Status` field.
+- To support new technologies such as IPv6.
+
+## Proposal
+
+### Custom Resource Definitions
+
+The new set of CRDs will be defined as part of a `v1alpha2` API.
+
+#### `Hardware`
+
+```go
+// Hardware is a logical representation of a machine that can execute Workflows.
+type Hardware struct {
+	HardwareSpec
```

Comment:
There is nothing in Kubernetes guidlines that says we must have a `Spec` or `Status` field. I'm wondering if we should just move all the `HardwareSpec` fields up to the `Hardware` object.

Another thread is talking about the possibility of adding a `Status` field which, if that happened, would necessitate the `HardwareSpec` but I don't think we have any intention of adding hardware status information any time soon so I suspect this would be fine.

Reply:
I am not quite sure that there is something written about any CRD should have a field called Spec, or Status, however, if you take a look at k8s crds you will find that, they do have such fields which makes other k8s users familiar with our CRD and know where to look whenever they need to adjust something.

With that being said, we can explore this path in the future, and see how those fields can be used during the lifecycle of those CRs.